File: PILA CON LISTAS/pilalistaE.js
class Node {
    constructor(value) {
        this.data = value;
        this.next = null;
    }
}

class Stack {
    constructor() {
        this.top = null;
    }
    
    push(num) {
        const newNode = new Node(num);
        newNode.next = this.top;
        this.top = newNode;
    }
    
    pop() {
        if (this.isEmpty()) {
            console.log("Stack Underflow");
            return -1;
        }
        
        const poppedValue = this.top.data;
        this.top = this.top.next;
        return poppedValue;
    }
    
    peek() {
        if (this.isEmpty()) {
            console.log("Stack is empty");
            return -1;
        }
        return this.top.data;
    }
    
    isEmpty() {
        return this.top === null;
    }
    
    display() {
        if (this.isEmpty()) {
            console.log("Stack is empty");
            return;
        }
        
        let current = this.top;
        let elements = [];
        while (current !== null) {
            elements.push(current.data);
            current = current.next;
        }
        console.log("Elementos en la pila: " + elements.join(" "));
    }
}

// Uso
const s = new Stack();
s.push(10);
s.push(20);
s.push(30);

console.log("Elemento Superior: " + s.peek());
console.log("Extrae elemento: " + s.pop());
console.log("Elemento Superior: " + s.peek());
s.display();

// Pruebas adicionales
console.log("\n--- Pruebas adicionales ---");
s.push(40);
s.push(50);
console.log("Elemento Superior después de nuevos push: " + s.peek());
s.display();

console.log("Extrae elemento: " + s.pop());
console.log("Extrae elemento: " + s.pop());
console.log("Extrae elemento: " + s.pop());
console.log("Intenta extraer de pila vacía: " + s.pop());
s.display();

File: PILA CON LISTAS/pilalistaE.cs
using System;

public class Node
{
    public int Data { get; set; }
    public Node Next { get; set; }
    
    public Node(int value)
    {
        Data = value;
        Next = null;
    }
}

public class Stack
{
    private Node top;
    
    public Stack()
    {
        top = null;
    }
    
    public void Push(int num)
    {
        Node newNode = new Node(num);
        newNode.Next = top;
        top = newNode;
    }
    
    public int Pop()
    {
        if (IsEmpty())
        {
            Console.WriteLine("Stack Underflow");
            return -1;
        }
        
        int poppedValue = top.Data;
        top = top.Next;
        return poppedValue;
    }
    
    public int Peek()
    {
        if (IsEmpty())
        {
            Console.WriteLine("Stack is empty");
            return -1;
        }
        return top.Data;
    }
    
    public bool IsEmpty()
    {
        return top == null;
    }
    
    public void Display()
    {
        if (IsEmpty())
        {
            Console.WriteLine("Stack is empty");
            return;
        }
        
        Node current = top;
        Console.Write("Elementos en la pila: ");
        while (current != null)
        {
            Console.Write(current.Data + " ");
            current = current.Next;
        }
        Console.WriteLine();
    }
}

class Program
{
    static void Main(string[] args)
    {
        Stack s = new Stack();
        s.Push(10);
        s.Push(20);
        s.Push(30);

        Console.WriteLine("Elemento Superior: " + s.Peek());
        Console.WriteLine("Extrae elemento: " + s.Pop());
        Console.WriteLine("Elemento Superior: " + s.Peek());
        s.Display();
    }
}